### ISSUE TITLE
dotnet target files sometimes have the wrong part of the string linked

### ISSUE THREAD

**Issue by xt0rted** [OctoLinker/OctoLinker#1062]
**Describe the bug**

If there are multiple instances of the text to be linked, and they're the same case on the same line, the wrong one can be linked. For the links that are in the wrong spot they're also including the character before & after the matched text.

The plugin that's running on this file is [`plugin-dotnet`](https://github.com/OctoLinker/OctoLinker/blob/dbaaaa5d845ed83198043201150753539210ee09/packages/plugin-dotnet/index.js).

**To Reproduce**

1. Go to https://gist.github.com/xt0rted/f6bf5f0d9c4bbb42f829cb8f0163957c#file-directory-build-targets
2. Add `octolinker-debug` to the body
3. View lines 4, 5, 16, and 17

**Expected behavior**

For `Include="NUnit"` to be linked, not `Version="$(NUnitVersion)"`.

**Additional context**

![image](https://user-images.githubusercontent.com/831974/95171477-ec35ef80-0783-11eb-9c14-219eed5d6edc.png)

**Comment by edavidaja** [comment]
Variation on the theme: if an R function is called with an identically named namespace operator, the function call and some following punctuation are linked rather than the package name, which precedes the two colons.

 ![image](https://user-images.githubusercontent.com/12927842/95793563-e2484b00-0cb3-11eb-889a-fbd31ae308ca.png)

**Comment by fregante** [comment]
Variations of this issue have been reported before: https://github.com/OctoLinker/OctoLinker/issues/618

It seems to me that the matching happens twice in 2 separate ways: once to detect the dependencies and once to linkify the content via `helper-insert-link`.

If the first matching happens correctly, why not just reuse the same matched DOM element instead of using `findAndReplaceDOMText`?

**Comment by stefanbuck** [comment]
Yes, matching happens twice as you describe. The first match operates on string representation of the code. This match returns an object like this

```json
{
  "endPos": 5,
  "endPosInBlob": 11,
  "lineNumber": 2,
  "startPos": 0,
  "startPosInBlob": 6,
  "values": [
    "hello-world",
  ],
},
```

Then `findAndReplaceDOMText` is used to wrap the match with an anchor tag. This part seems to be buggy and I started looking into this last weekend, but wasn't able to finish it. 

We use this `findAndReplaceDOMText` because a match may span across multiple child nodes like in this example.


```html
<td id="LC19" class="blob-code blob-code-inner js-file-line">
    <span class="pl-k">use</span> <span class="pl-v">Illuminate</span>\<span class="pl-v">Contracts</span>\<span class="pl-v">Container</span>\<span class="pl-v">BindingResolutionException</span>;
</td>
```

Also in a diff view, one or more additional <span>'s maybe surrounding the match 

![image](https://user-images.githubusercontent.com/1393946/112393496-73a7f880-8cfb-11eb-9055-c5d7a32cbc26.png)

However, as mentioned before, I started looking into this but I need more time to investigate this further. This part is the most complicated bit in OctoLinker, but it is necessary to decouple the code from the underlaying DOM as much as possible. In the early days OctoLinker broke a few times because GitHub updates classnames I used to find and wrap matches. 

@fregante I'm curious to know to how [refined-github](https://github.com/sindresorhus/refined-github) is dealing with this problem.

**Comment by fregante** [comment]
I think I suggested it before: https://github.com/fregante/zip-text-node + a custom script to linkify text, for example https://github.com/sindresorhus/linkify-issues

The latter receives plain text and returns a DocumentFragment with the same text, but linkified. Then this is passed to zip-text-nodes so that this new Anchor element is merged into the original DOM, by basically diffing it.

This works well on titles or single lines but I don’t know how well it could work on whole files. Probably you could linkify the whole file but then only pass the single line to be *zipped* together.
